### PR TITLE
Upgrade to RxJava 2

### PR DESCRIPTION
**Pull request by dlew**
The goal here is just to start the big transition, I'm sure there are
lots of little polishes/improvements we can do once this is in.

Some of the bigger changes:

- Removed dependency on RxBinding since there is no RxJava 2-based
  RxBinding yet. Implementing our own ViewDetachesOnSubscribe was
  not that bad.

- Rolled all Transformers into a single LifecycleTransformer, now that
  we can do so (due to composition-friendly interfaces).

### CODE REVIEW

File: rxlifecycle-kotlin/src/main/java/com/trello/rxlifecycle/kotlin/rxlifecycle.kt [hunk 1]
```diff
@@ -31,19 +31,19 @@ fun <T> Observable<T>.bindToLifecycle(view: View): Observable<T>
...
+        = this.compose(provider.bindToLifecycle<T>())
 
 fun <T, E> Single<T>.bindUntilEvent(provider: LifecycleProvider<E>, event: E): Single<T>
-        = this.compose(provider.bindUntilEvent<T>(event).forSingle<T>())
+        = this.compose(provider.bindUntilEvent<T>(event))
 
 fun <T> Single<T>.bindToLifecycle(view: View): Single<T>
-        = this.compose(RxLifecycleAndroid.bindView<T>(view).forSingle<T>())
\ No newline at end of file
+        = this.compose(RxLifecycleAndroid.bindView<T>(view))
```
Comment (vanniktech) [inline rxlifecycle-kotlin/src/main/java/com/trello/rxlifecycle/kotlin/rxlifecycle.kt]: Missing extension methods for Maybe + Flowable. So far I also didn't see that support was added for those. This is just meant as a reminder.

File: rxlifecycle/src/main/java/com/trello/rxlifecycle/RxLifecycle.java [hunk 2]
```diff
@@ -47,28 +48,28 @@ private RxLifecycle() {
...
+     * Use with {@link Observable#compose(ObservableTransformer)}:
      * {@code source.compose(RxLifecycle.bind(lifecycle)).subscribe()}
      * <p>
      * This helper automatically determines (based on the lifecycle sequence itself) when the source
      * should stop emitting items. Note that for this method, it assumes <em>any</em> event
      * emitted by the given lifecycle indicates that the lifecycle is over.
      *
      * @param lifecycle the lifecycle sequence
-     * @return a reusable {@link Observable.Transformer} that unsubscribes the source whenever the lifecycle emits
+     * @return a reusable {@link ObservableTransformer} that unsubscribes the source whenever the lifecycle emits
```
Comment (vanniktech) [inline rxlifecycle/src/main/java/com/trello/rxlifecycle/RxLifecycle.java]: it returns a `LifecycleTransformer`

File: rxlifecycle/src/main/java/com/trello/rxlifecycle/RxLifecycle.java [hunk 3]
```diff
@@ -33,12 +34,12 @@ private RxLifecycle() {
...
      * When the lifecycle event occurs, the source will cease to emit any notifications.
      * <p>
-     * Use with {@link Observable#compose(Observable.Transformer)}:
+     * Use with {@link Observable#compose(ObservableTransformer)}:
      * {@code source.compose(RxLifecycle.bindUntilEvent(lifecycle, ActivityEvent.STOP)).subscribe()}
      *
      * @param lifecycle the lifecycle sequence
      * @param event the event which should conclude notifications from the source
-     * @return a reusable {@link Observable.Transformer} that unsubscribes the source at the specified event
+     * @return a reusable {@link ObservableTransformer} that unsubscribes the source at the specified event
```
Comment (vanniktech) [inline rxlifecycle/src/main/java/com/trello/rxlifecycle/RxLifecycle.java]: it returns `LifecycleTransformer` that implements `ObservableTransformer` plus the one for Single + Completable

File: rxlifecycle/src/main/java/com/trello/rxlifecycle/RxLifecycle.java [hunk 4]
```diff
@@ -83,16 +84,16 @@ private RxLifecycle() {
      *
      * @param lifecycle the lifecycle sequence
      * @param correspondingEvents a function which tells the source when to unsubscribe
-     * @return a reusable {@link Observable.Transformer} that unsubscribes the source during the Fragment lifecycle
+     * @return a reusable {@link ObservableTransformer} that unsubscribes the source during the Fragment lifecycle
      */
     @Nonnull
     @CheckReturnValue
     public static <T, R> LifecycleTransformer<T> bind(@Nonnull Observable<R> lifecycle,
```
Comment (vanniktech) [inline rxlifecycle/src/main/java/com/trello/rxlifecycle/RxLifecycle.java]: same

File: rxlifecycle/src/main/java/com/trello/rxlifecycle/UntilLifecycleTransformer.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Override
+    public SingleSource<T> apply(Single<T> upstream) {
+        return upstream.takeUntil(lifecycle.take(1).singleOrError());
+    }
+
+    @Override
+    public CompletableSource apply(Completable upstream) {
+        return Completable.amb(
```
Comment (vanniktech) [inline rxlifecycle/src/main/java/com/trello/rxlifecycle/UntilLifecycleTransformer.java]: maybe consider using Completable.ambArray here

File: rxlifecycle/src/main/java/com/trello/rxlifecycle/UntilEventTransformer.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Override
+    public SingleSource<T> apply(Single<T> upstream) {
+        return upstream.takeUntil(takeUntilEvent(lifecycle, event).singleOrError());
+    }
+
+    @Override
+    public CompletableSource apply(Completable upstream) {
+        return Completable.amb(
```
Comment (vanniktech) [inline rxlifecycle/src/main/java/com/trello/rxlifecycle/UntilEventTransformer.java]: maybe consider using Completable.ambArray here

File: rxlifecycle/src/main/java/com/trello/rxlifecycle/UntilCorrespondingEventTransformer.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    @Override
+    public SingleSource<T> apply(Single<T> upstream) {
+        return upstream.takeUntil(takeUntilCorrespondingEvent(sharedLifecycle, correspondingEvents).singleOrError());
+    }
+
+    @Override
+    public CompletableSource apply(Completable upstream) {
+        return Completable.amb(
```
Comment (vanniktech) [inline rxlifecycle/src/main/java/com/trello/rxlifecycle/UntilCorrespondingEventTransformer.java]: maybe consider using Completable.ambArray here

File: rxlifecycle/src/test/java/com/trello/rxlifecycle/UntilLifecycleTransformerSingleTest.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Test
+    public void noEvent() {
+        TestObserver<String> testObserver = Single.just("1")
+            .delay(1, TimeUnit.MILLISECONDS, testScheduler)
+            .compose(new UntilLifecycleTransformer<String, String>(lifecycle))
+            .test();
+
+        testObserver.assertNoValues();
```
Comment (vanniktech) [inline rxlifecycle/src/test/java/com/trello/rxlifecycle/UntilLifecycleTransformerSingleTest.java]: nice 👍

File: rxlifecycle-kotlin/src/main/java/com/trello/rxlifecycle/kotlin/rxlifecycle.kt (hunk 1 above)
Comment (dlew) [inline rxlifecycle-kotlin/src/main/java/com/trello/rxlifecycle/kotlin/rxlifecycle.kt]: See #163 and #164

**Pull request merged**